McCorrections/src/LeptonWeights.cpp

```diff
@@ -65,53 +65,70 @@ LeptonWeights::LeptonWeights(const std::string& electron_idIsoInput, const std::
-                                                   static_cast<int>(eventInfo.GetLeg(2)->gen_match()));
-        return electronSF.GetIdIsoSF(eventInfo.GetLeg(1).GetMomentum()) * tau_weight;
+    std::map<Period ,std::string>  period_label = { { Period::Run2016, "2016Legacy" },
+                                                     { Period::Run2017, "2017ReReco" },
+                                                     { Period::Run2018, "2018ReReco" } };
+    std::string s = "by";
+    std::string::size_type i = ToString(discr).find(s);
+
+    tau_sf_providers[discr][wp] = std::make_shared<TauIDSFTool>(period_label.at(period),
+                                ToString(discr).erase(i, s.length()), ToString(wp), false, false);
+    return tau_sf_providers[discr][wp];
+}
+
+double LeptonWeights::GetLegIdIsoWeight(LepCandidate leg, DiscriminatorWP VSe_wp, DiscriminatorWP VSmu_wp,
+                                        DiscriminatorWP VSjet_wp,
+                                        UncertaintySource /*unc_source*/, UncertaintyScale /*unc_scale*/)
+{
+    std::map<LegType, std::pair<TauIdDiscriminator,DiscriminatorWP>> leg_wp_disc = {
+         { LegType::e, { TauIdDiscriminator::byDeepTau2017v2p1VSe, VSe_wp } },
+         { LegType::mu, { TauIdDiscriminator::byDeepTau2017v2p1VSmu, VSmu_wp } },
+         { LegType::tau, { TauIdDiscriminator::byDeepTau2017v2p1VSjet, VSjet_wp } }
+     };
+
+
+    auto tauIdWeight = GetTauIdProvider(leg_wp_disc.at(leg->leg_type()).first, leg_wp_disc.at(leg->leg_type()).second);
+
+    if(leg->leg_type() == LegType::e){
+        double tau_id_weight = tauIdWeight->getSFvsEta(leg.GetMomentum().eta(), static_cast<int>(leg->gen_match()));
+        return electronSF.GetIdIsoSF(leg.GetMomentum()) * tau_id_weight;
     }
-    else if(channel == Channel::MuTau) {
-        double tau_weight = tauIdWeight->getSFvsPT(eventInfo.GetLeg(2).GetMomentum().pt(),
-                                                   static_cast<int>(eventInfo.GetLeg(2)->gen_match()));
-        return muonSF.GetIdIsoSF(eventInfo.GetLeg(1).GetMomentum()) * tau_weight;
+    else if(leg->leg_type() == LegType::mu){
+        double tau_id_weight = tauIdWeight->getSFvsEta(leg.GetMomentum().eta(), static_cast<int>(leg->gen_match()));
+        return  muonSF.GetIdIsoSF(leg.GetMomentum()) * tau_id_weight;
     }
-    else if(channel == Channel::TauTau) {
-        double tau_weight = tauIdWeight->getSFvsPT(eventInfo.GetLeg(1).GetMomentum().pt(),
-                                                   static_cast<int>(eventInfo.GetLeg(1)->gen_match()))
-                          * tauIdWeight->getSFvsPT(eventInfo.GetLeg(2).GetMomentum().pt(),
-                                                   static_cast<int>(eventInfo.GetLeg(2)->gen_match()));
-        return tau_weight;
+    else if(leg->leg_type() == LegType::tau){
+        double tau_id_weight = tauIdWeight->getSFvsPT(leg.GetMomentum().pt(), static_cast<int>(leg->gen_match()));
```

Comment:
depends on tau gen match and for true taus whatever it is dm or pt dependent SFs